src/catalog-debug/src/main.rs

```diff
@@ -0,0 +1,511 @@
+#![warn(clippy::unnecessary_lazy_evaluations)]
+#![warn(clippy::partialeq_ne_impl)]
+#![warn(clippy::redundant_field_names)]
+#![warn(clippy::transmutes_expressible_as_ptr_casts)]
+#![warn(clippy::unused_async)]
+#![warn(clippy::disallowed_methods)]
+#![warn(clippy::disallowed_macros)]
+#![warn(clippy::disallowed_types)]
+#![warn(clippy::from_over_into)]
+// END LINT CONFIG
+
+//! Debug utility for stashes.
+
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::fs::File;
+use std::io::{self, Write};
+use std::path::PathBuf;
+use std::process;
+use std::str::FromStr;
+use std::sync::{Arc, Mutex};
+
+use anyhow::Context;
+use clap::clap_derive::ArgEnum;
+use clap::Parser;
+use mz_adapter::catalog::{Catalog, ClusterReplicaSizeMap, Config};
+use mz_build_info::{build_info, BuildInfo};
+use mz_catalog::durable::debug::{
+    AuditLogCollection, ClusterCollection, ClusterIntrospectionSourceIndexCollection,
+    ClusterReplicaCollection, Collection, CollectionTrace, CollectionType, CommentCollection,
+    ConfigCollection, DatabaseCollection, DebugCatalogState, DefaultPrivilegeCollection,
+    IdAllocatorCollection, ItemCollection, RoleCollection, SchemaCollection, SettingCollection,
+    StorageUsageCollection, SystemConfigurationCollection, SystemItemMappingCollection,
+    SystemPrivilegeCollection, TimestampCollection, Trace,
+};
+use mz_catalog::durable::{
+    persist_backed_catalog_state, stash_backed_catalog_state, BootstrapArgs,
+    OpenableDurableCatalogState, StashConfig,
+};
+use mz_ore::cli::{self, CliConfig};
+use mz_ore::error::ErrorExt;
+use mz_ore::metrics::MetricsRegistry;
+use mz_ore::now::SYSTEM_TIME;
+use mz_persist_client::cache::PersistClientCache;
+use mz_persist_client::cfg::PersistConfig;
+use mz_persist_client::rpc::PubSubClientConnection;
+use mz_persist_client::PersistLocation;
+use mz_secrets::InMemorySecretsController;
+use mz_sql::catalog::EnvironmentId;
+use mz_sql::session::vars::ConnectionCounter;
+use mz_stash::StashFactory;
+use once_cell::sync::Lazy;
+use serde::Serialize;
+use url::Url;
+use uuid::Uuid;
+
+pub const BUILD_INFO: BuildInfo = build_info!();
+pub static VERSION: Lazy<String> = Lazy::new(|| BUILD_INFO.human_version());
+
+#[derive(Parser, Debug)]
+#[clap(name = "catalog", next_line_help = true, version = VERSION.as_str())]
+pub struct Args {
+    #[clap(long, arg_enum)]
+    store: CatalogKind,
+
+    // === Stash options. ===
+    /// The PostgreSQL URL for the adapter stash.
+    #[clap(long, env = "ADAPTER_STASH_URL", required_if_eq("store", "stash"))]
+    postgres_url: Option<String>,
+
+    // === Persist options. ===
+    /// The organization ID of the environment.
+    #[clap(long, env = "ORG_ID", required_if_eq("store", "persist"))]
+    organization_id: Option<Uuid>,
+    /// Where the persist library should store its blob data.
+    #[clap(long, env = "PERSIST_BLOB_URL", required_if_eq("store", "persist"))]
+    persist_blob_url: Option<Url>,
```

Comment:
> The debug tool shouldn't have to write any data to Persist

I don't think that's entirely true. The `edit` and `delete` commands will have to write data to persist to have an effect. Only the `dump` and `upgrade-check` commands are read-only.

Reply:
If you like, persist has some `ReadOnly` s3/crdb wrappers for the read-only cmd in persistcli that we could expose more widely. We mostly have them for a --no-commit/--commit sort of thing on our read-write commands, so we don't accidentally  introduce a write in the --no-commit version, but if you're stressed about `dump` and `upgrade-check` accidentally writing things, this might help https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/cli/args.rs#L122-L159

Reply:
@jkosh44 yeah you're totally right, took me a minute to get my brain warmed up this morning, ignore me!

Reply:
No worries, I'll probably hold off on the read-only persist stuff for now. Maybe we can put that into the catalog impl itself when calling `open_read_only` or `open_savepoint`.